doc/refman/RefMan-ssr.tex

```diff
@@ -0,0 +1,4687 @@
+Explicit pattern testing also largely subsumes the generalization of
+the \ssrC{if} construct to all binary datatypes; compare:
+
+\ssrC{if} \ssrN{term} \ssrC{is inl _ then} \ssrN[l]{term} \ssrC{else} \ssrN[r]{term}
+
+and:
+
+\ssrC{if} \ssrN{term} \ssrC{then} \ssrN[l]{term} \ssrC{else} \ssrN[r]{term}
+
+The latter appears to be marginally shorter, but it is quite
+ambiguous, and indeed often
+requires an explicit annotation term : \ssrC{\{_\}+\{_\}} to type-check,
+which evens the character count.
+
+Therefore, \ssr{} restricts by default the condition of a plain \ssrC{if}
+construct to the standard \ssrC{bool} type; this avoids spurious type
+annotations, e.g., in:
+\begin{lstlisting}
+  Definition |*orb*| b1 b2 := if b1 then true else b2.
+\end{lstlisting}
+As pointed out in section 1.2, this restriction can be removed with
+the command:
+\begin{lstlisting}
+  Close Scope boolean_if_scope.
+\end{lstlisting}
+Like \ssrC{let:} above, the \ssrC{if} \ssrN{term} \ssrC{is} \ssrN{pattern}
+\ssrC{else} \ssrN{term} construct
+supports the dependent \ssrC{match} annotations:
+
+\ssrC{if} \ssrN[1]{term} \ssrC{is} \ssrN[1]{pattern} \ssrC{as} \ssrN{ident} \ssrC{in} \ssrN[2]{pattern} \ssrC{return} \ssrN[2]{term} \ssrC{then} \ssrN[3]{term} \ssrC{else} \ssrN[4]{term}
+
+As in \ssrC{let:} the variable \ssrN{ident} (and those in
+the type pattern \ssrN[2]{pattern}) are bound in \ssrN[2]{term}; \ssrN{ident} is
+also bound in \ssrN[3]{term} (but not in \ssrN[4]{term}), while the
+variables in \ssrN[1]{pattern} are bound only in \ssrN[3]{term}.
+
+\noindent
+Another variant allows to treat the else case first:
+
+\ssrC{if} \ssrN[1]{term} \ssrC{isn't} \ssrN[1]{pattern} \ssrC{then} \ssrN[2]{term} \ssrC{else} \ssrN[3]{term}
+
+Note that \ssrN[1]{pattern} eventually binds variables in \ssrN[3]{term}
+and not \ssrN[2]{term}.
+
+\subsection{Parametric polymorphism}\label{ssec:parampoly}
+Unlike ML, polymorphism in core Gallina is explicit: the type
+parameters of polymorphic  functions must be declared explicitly, and
+supplied at each point of use. However, \Coq{} provides two features
+to suppress redundant parameters:
+\begin{itemize}
+\item Sections are used to provide (possibly implicit) parameters for
+  a set of definitions.
+\item Implicit arguments declarations are used to tell \Coq{} to use
+  type inference to deduce some parameters from the context at each
+  point of call.
+\end{itemize}
+The combination of these features provides a fairly good emulation of ML-style
+polymorphism, but unfortunately this emulation breaks down for
+higher-order programming. Implicit arguments are indeed not inferred
+at all points of use, but only at
+points of call, leading to  expressions such as
+\begin{lstlisting}
+  Definition |*all_null*| (s : list T) := all (@null T) s.
+\end{lstlisting}
+Unfortunately, such higher-order expressions are quite frequent in
+representation functions, especially those which use \Coq{}'s
+\ssrC{Structure}s to emulate Haskell type classes.
```

Comment:
Clearly this sentence needs some updating since type classes are available in Coq (and BTW it's not just Structure, which is equivalent to Record, but Canonical Structure, right?)